superset-frontend/src/explore/components/ControlPanelsContainer.tsx

```diff
@@ -190,57 +174,79 @@ function getState(
-      this.setState(getState(this.props));
-    }
-  }
+  const prevDatasource = usePrevious(props.exploreState.datasource);
 
-  // required for an Antd bug that would otherwise malfunction re-rendering
-  // a collapsed panel after changing the datasource or viz type
-  UNSAFE_componentWillReceiveProps(nextProps: ControlPanelsContainerProps) {
+  const [expandedQuerySections, setExpandedQuerySections] = useState<string[]>(
+    [],
+  );
+  const [expandedCustomizeSections, setExpandedCustomizeSections] = useState<
+    string[]
+  >([]);
+  const [querySections, setQuerySections] = useState<
+    ControlPanelSectionConfig[]
+  >([]);
+  const [customizeSections, setCustomizeSections] = useState<
+    ControlPanelSectionConfig[]
+  >([]);
+  const [showDatasourceAlert, setShowDatasourceAlert] = useState(false);
+
+  useEffect(() => {
     if (
-      this.props.form_data.datasource !== nextProps.form_data.datasource ||
-      this.props.form_data.viz_type !== nextProps.form_data.viz_type
+      prevDatasource &&
+      (props.exploreState.datasource?.id !== prevDatasource.id ||
+        props.exploreState.datasource?.type !== prevDatasource.type)
     ) {
-      this.setState({ loading: true });
+      setShowDatasourceAlert(true);
     }
-  }
+  }, [
+    props.exploreState.datasource?.id,
+    props.exploreState.datasource?.type,
+    prevDatasource,
+  ]);
+
+  useEffect(() => {
+    const {
+      expandedQuerySections: newExpandedQuerySections,
+      expandedCustomizeSections: newExpandedCustomizeSections,
+      querySections: newQuerySections,
+      customizeSections: newCustomizeSections,
+    } = getState(
+      props.form_data.viz_type,
+      props.exploreState.datasource,
+      props.datasource_type,
+    );
+    setExpandedQuerySections(newExpandedQuerySections);
+    setExpandedCustomizeSections(newExpandedCustomizeSections);
+    setQuerySections(newQuerySections);
+    setCustomizeSections(newCustomizeSections);
+  }, [props.form_data.datasource, props.form_data.viz_type]);
+
+  const resetTransferredControls = useCallback(() => {
+    ensureIsArray(props.exploreState.controlsTransferred).forEach(controlName =>
+      props.actions.setControlValue(
+        controlName,
+        props.controls[controlName].default,
+      ),
+    );
+  }, [props.actions, props.exploreState.controlsTransferred, props.controls]);
 
-  componentDidMount() {
-    this.setState(getState(this.props));
-  }
+  const handleClearFormClick = useCallback(() => {
+    resetTransferredControls();
+    setShowDatasourceAlert(false);
+  }, [resetTransferredControls]);
+
+  const handleContinueClick = useCallback(() => {
+    setShowDatasourceAlert(false);
+  }, []);
 
-  renderControl({ name, config }: CustomControlItem) {
-    const { actions, controls, chart, exploreState } = this.props;
+  const renderControl = ({ name, config }: CustomControlItem) => {
+    const { controls, chart } = props;
```

Comment:
It seems we're missing `exploreState` here (needed down on line 257):
```suggestion
    const { controls, chart, exploreState } = props;
```

Reply:
🤦 I moved `exploreState` into `ControlPanelsContainer`, but then I changed my mind and moved it back to props. For some reason, linter didn't mark this line with "variable not found" error. Thanks for finding that!